// @keystatic/core/component-blocks/dist/keystatic-core-component-blocks.react-server.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var api = require('../../dist/api-2733d27f.react-server.cjs.js');
var index = require('../../dist/index-63269275.react-server.cjs.js');
require('../../dist/empty-field-ui-563fc621.react-server.cjs.js');
require('react/jsx-runtime');
require('emery');
require('@sindresorhus/slugify');
require('@braintree/sanitize-url');
require('@markdoc/markdoc');
require('slate');
require('emery/assertions');
require('js-base64');
require('@emotion/weak-memoize');

function CloudImagePreview() {}

const cloudImageSchema = {
  src: index.text({
    label: 'URL',
    validation: {
      length: {
        min: 1
      }
    }
  }),
  alt: index.text({
    label: 'Alt text'
  }),
  height: api.integer({
    label: 'Height'
  }),
  width: api.integer({
    label: 'Width'
  })
};

/** @deprecated Experimental */
function cloudImage(args) {
  return api.component({
    label: args.label,
    schema: cloudImageSchema,
    preview: CloudImagePreview,
    chromeless: true,
    toolbar: null
  });
}

exports.cloudImage = cloudImage;
