// @keystatic/core/reader/dist/keystatic-core-reader.react-server.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('react/jsx-runtime');
var index = require('../../dist/index-63269275.react-server.cjs.js');
require('@sindresorhus/slugify');
require('@braintree/sanitize-url');
var fs = require('fs/promises');
var path = require('path');
var requiredFiles = require('../../dist/required-files-99a2d1f2.react-server.cjs.js');
var React = require('react');
require('@markdoc/markdoc');
require('slate');
require('emery/assertions');
require('emery');
require('js-base64');
require('../../dist/empty-field-ui-563fc621.react-server.cjs.js');
require('@emotion/weak-memoize');
require('js-yaml');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var fs__default = /*#__PURE__*/_interopDefault(fs);
var path__default = /*#__PURE__*/_interopDefault(path);
var React__namespace = /*#__PURE__*/_interopNamespace(React);

function cache$1(func) {
  return func;
}

// we conditionally using it since it's not actually in stable react releases yet
// (though it should be unnecessary since this file is only imported in react-server environments anyway)
// it's a function because some tools try to be smart with accessing things on namespace imports
// and error at build time if you try to read an export that doesn't exist on a namespace object
function getCache(react) {
  var _react$cache;
  return (_react$cache = react.cache) !== null && _react$cache !== void 0 ? _react$cache : cache$1;
}
const cache = getCache(React__namespace);

async function getAllEntries(root, prefix) {
  return (await Promise.all((await fs__default["default"].readdir(path__default["default"].join(root, prefix), {
    withFileTypes: true
  }).catch(err => {
    if (err.code === 'ENOENT') {
      return [];
    }
    throw err;
  })).map(async dirent => {
    const name = `${prefix}${dirent.name}`;
    const entry = {
      entry: dirent,
      name
    };
    if (dirent.isDirectory()) {
      return [entry, ...(await getAllEntries(root, `${name}/`))];
    }
    if (dirent.isFile()) {
      return entry;
    }
    return [];
  }))).flat();
}
const listCollection = cache(async function listCollection(repoPath, collectionPath, glob, formatInfo, extension) {
  const entries = glob === '*' ? (await fs__default["default"].readdir(path__default["default"].join(repoPath, collectionPath), {
    withFileTypes: true
  }).catch(err => {
    if (err.code === 'ENOENT') {
      return [];
    }
    throw err;
  })).map(x => ({
    entry: x,
    name: x.name
  })) : await getAllEntries(path__default["default"].join(repoPath, collectionPath), '');
  return (await Promise.all(entries.map(async x => {
    if (formatInfo.dataLocation === 'index') {
      if (!x.entry.isDirectory()) return [];
      try {
        await fs__default["default"].stat(path__default["default"].join(repoPath, index.getEntryDataFilepath(`${collectionPath}/${x.name}`, formatInfo)));
        return [x.name];
      } catch (err) {
        if (err.code === 'ENOENT') {
          return [];
        }
        throw err;
      }
    } else {
      if (!x.entry.isFile() || !x.name.endsWith(extension)) return [];
      return [x.name.slice(0, -extension.length)];
    }
  }))).flat();
});
function collectionReader(repoPath, collection, config) {
  const formatInfo = index.getCollectionFormat(config, collection);
  const collectionPath = index.getCollectionPath(config, collection);
  const collectionConfig = config.collections[collection];
  const schema = index.object(collectionConfig.schema);
  const glob = index.getSlugGlobForCollection(config, collection);
  const extension = index.getDataFileExtension(formatInfo);
  const read = (slug, ...args) => {
    var _args$;
    return readItem(schema, formatInfo, index.getCollectionItemPath(config, collection, slug), repoPath, (_args$ = args[0]) === null || _args$ === void 0 ? void 0 : _args$.resolveLinkedFiles, `"${slug}" in collection "${collection}"`, slug, collectionConfig.slugField, glob);
  };
  const list = () => listCollection(repoPath, collectionPath, glob, formatInfo, extension);
  return {
    read,
    readOrThrow: async (...args) => {
      const entry = await read(...args);
      if (entry === null) {
        throw new Error(`Entry "${args[0]}" not found in collection "${collection}"`);
      }
      return entry;
    },
    // TODO: this could drop the fs.stat call that list does for each item
    // since we just immediately read it
    all: async (...args) => {
      const slugs = await list();
      return (await Promise.all(slugs.map(async slug => {
        const entry = await read(slug, args[0]);
        if (entry === null) return [];
        return [{
          slug,
          entry
        }];
      }))).flat();
    },
    list
  };
}
const readItem = cache(async function readItem(rootSchema, formatInfo, itemDir, repoPath, resolveLinkedFiles, debugReference, ...slugInfo) {
  let dataFile;
  try {
    dataFile = await fs__default["default"].readFile(path__default["default"].resolve(repoPath, index.getEntryDataFilepath(itemDir, formatInfo)));
  } catch (err) {
    if (err.code === 'ENOENT') {
      return null;
    }
    throw err;
  }
  const {
    loaded,
    extraFakeFile
  } = requiredFiles.loadDataFile(dataFile, formatInfo);
  const contentFieldPathsToEagerlyResolve = resolveLinkedFiles ? [] : undefined;
  let validated;
  try {
    validated = requiredFiles.parseProps(rootSchema, loaded, [], [], (schema, value, path, pathWithArrayFieldSlugs) => {
      if (schema.formKind === 'asset') {
        return schema.reader.parse(value);
      }
      if (schema.formKind === 'content') {
        contentFieldPathsToEagerlyResolve === null || contentFieldPathsToEagerlyResolve === void 0 || contentFieldPathsToEagerlyResolve.push(path);
        return async () => {
          let content;
          const filename = pathWithArrayFieldSlugs.join('/') + schema.contentExtension;
          if (filename === (extraFakeFile === null || extraFakeFile === void 0 ? void 0 : extraFakeFile.path)) {
            content = extraFakeFile.contents;
          } else {
            content = await fs__default["default"].readFile(path__default["default"].resolve(repoPath, `${itemDir}/${filename}`)).catch(x => {
              if (x.code === 'ENOENT') return undefined;
              throw x;
            });
          }
          return schema.reader.parse(value, {
            content
          });
        };
      }
      if (path.length === 1 && slugInfo[0] !== undefined) {
        const [slug, slugField, glob] = slugInfo;
        if (path[0] === slugField) {
          if (schema.formKind !== 'slug') {
            throw new Error(`Slug field ${slugInfo[1]} is not a slug field`);
          }
          return schema.reader.parseWithSlug(value, {
            slug,
            glob
          });
        }
      }
      return schema.reader.parse(value);
    }, true);
    if (contentFieldPathsToEagerlyResolve !== null && contentFieldPathsToEagerlyResolve !== void 0 && contentFieldPathsToEagerlyResolve.length) {
      await Promise.all(contentFieldPathsToEagerlyResolve.map(async path => {
        const parentValue = index.getValueAtPropPath(validated, path.slice(0, -1));
        const keyOnParent = path[path.length - 1];
        const originalValue = parentValue[keyOnParent];
        parentValue[keyOnParent] = await originalValue();
      }));
    }
  } catch (err) {
    const formatted = requiredFiles.formatFormDataError(err);
    throw new Error(`Invalid data for ${debugReference}:\n${formatted}`);
  }
  return validated;
});
function singletonReader(repoPath, singleton, config) {
  const formatInfo = index.getSingletonFormat(config, singleton);
  const singletonPath = index.getSingletonPath(config, singleton);
  const schema = index.object(config.singletons[singleton].schema);
  const read = (...args) => {
    var _args$2;
    return readItem(schema, formatInfo, singletonPath, repoPath, (_args$2 = args[0]) === null || _args$2 === void 0 ? void 0 : _args$2.resolveLinkedFiles, `singleton "${singleton}"`, undefined);
  };
  return {
    read,
    readOrThrow: async (...opts) => {
      const entry = await read(...opts);
      if (entry === null) {
        throw new Error(`Singleton "${singleton}" not found`);
      }
      return entry;
    }
  };
}
function createReader(repoPath, config) {
  return {
    collections: Object.fromEntries(Object.keys(config.collections || {}).map(key => [key, collectionReader(repoPath, key, config)])),
    singletons: Object.fromEntries(Object.keys(config.singletons || {}).map(key => [key, singletonReader(repoPath, key, config)])),
    repoPath,
    config
  };
}

exports.createReader = createReader;
