// @keystatic/next/ui/app/dist/keystatic-next-ui-app.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var navigation = require('next/navigation');
var Link = require('next/link');
var react = require('react');
var ui = require('@keystatic/core/ui');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Link__default = /*#__PURE__*/_interopDefault(Link);

let _isClient = false;
function useIsClient() {
  const [isClient, setIsClient] = react.useState(_isClient);
  react.useEffect(() => {
    _isClient = true;
    setIsClient(true);
  }, []);
  return isClient;
}
function makePage(config) {
  return function Page() {
    const isClient = useIsClient();
    const router = navigation.useRouter();
    const pathname = navigation.usePathname();
    let href = pathname;
    const searchParams = navigation.useSearchParams().toString();
    if (searchParams) {
      href += `?${searchParams}`;
    }
    const keystaticRouter = react.useMemo(() => {
      const replaced = pathname.replace(/^\/keystatic\/?/, '');
      const params = replaced === '' ? [] : replaced.split('/').map(decodeURIComponent);
      return {
        href,
        params,
        push: async path => {
          router.push(path);
        },
        replace: async path => {
          router.replace(path);
        }
      };
    }, [href, router, pathname]);
    if (!isClient) return null;
    return /*#__PURE__*/jsxRuntime.jsx(ui.Keystatic, {
      router: keystaticRouter,
      config: config,
      link: Link__default["default"],
      appSlug: appSlug
    });
  };
}
const appSlug = {
  envName: 'NEXT_PUBLIC_KEYSTATIC_GITHUB_APP_SLUG',
  value: process.env.NEXT_PUBLIC_KEYSTATIC_GITHUB_APP_SLUG
};

exports.makePage = makePage;
