// @keystatic/next/ui/pages/dist/keystatic-next-ui-pages.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var router = require('next/router');
var Link = require('next/link');
var react = require('react');
var ui = require('@keystatic/core/ui');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Link__default = /*#__PURE__*/_interopDefault(Link);

function makePage(config) {
  return function Page() {
    const router$1 = router.useRouter();
    const keystaticRouter = react.useMemo(() => {
      var _ref, _router$query$params;
      if (!router$1.isReady) return null;
      const params = (_ref = (_router$query$params = router$1.query.params) !== null && _router$query$params !== void 0 ? _router$query$params : router$1.query.rest) !== null && _ref !== void 0 ? _ref : [];
      return {
        href: router$1.asPath,
        params,
        push: path => {
          router$1.push(path);
        },
        replace: path => {
          router$1.replace(path);
        }
      };
    }, [router$1]);
    if (!keystaticRouter) return null;
    return /*#__PURE__*/jsxRuntime.jsx(ui.Keystatic, {
      router: keystaticRouter,
      config: config,
      link: Link__default["default"],
      appSlug: appSlug
    });
  };
}
const appSlug = {
  envName: 'NEXT_PUBLIC_KEYSTATIC_GITHUB_APP_SLUG',
  value: process.env.NEXT_PUBLIC_KEYSTATIC_GITHUB_APP_SLUG
};

exports.makePage = makePage;
