// @keystatic/next/ui/pages/dist/keystatic-next-ui-pages.esm.js
import { useRouter } from 'next/router';
import Link from 'next/link';
import { useMemo } from 'react';
import { Keystatic } from '@keystatic/core/ui';
import { jsx } from 'react/jsx-runtime';

function makePage(config) {
  return function Page() {
    const router = useRouter();
    const keystaticRouter = useMemo(() => {
      var _ref, _router$query$params;
      if (!router.isReady) return null;
      const params = (_ref = (_router$query$params = router.query.params) !== null && _router$query$params !== void 0 ? _router$query$params : router.query.rest) !== null && _ref !== void 0 ? _ref : [];
      return {
        href: router.asPath,
        params,
        push: path => {
          router.push(path);
        },
        replace: path => {
          router.replace(path);
        }
      };
    }, [router]);
    if (!keystaticRouter) return null;
    return /*#__PURE__*/jsx(Keystatic, {
      router: keystaticRouter,
      config: config,
      link: Link,
      appSlug: appSlug
    });
  };
}
const appSlug = {
  envName: 'NEXT_PUBLIC_KEYSTATIC_GITHUB_APP_SLUG',
  value: process.env.NEXT_PUBLIC_KEYSTATIC_GITHUB_APP_SLUG
};

export { makePage };
