// @keystar/ui/next/dist/keystar-ui-next.esm.js
export { RootVoussoirProvider } from '../../dist/RootVoussoirProvider-89a6abba.esm.js';
export { NextRootProvider } from '../../dist/NextRootProvider-e938d779.esm.js';
import { jsx } from 'react/jsx-runtime';
export { useRootColorScheme } from '../../dist/useRootColorScheme-c64055b9.esm.js';

const script = `
let classList = document.documentElement.classList;
let storedPreference = localStorage.getItem('keystatic-root-color-scheme');
let schemeClasses = [...classList].filter((name) => name.includes('scheme'));

if (storedPreference === 'dark') {
  classList.remove(schemeClasses);
  classList.add('ksv-scheme--dark');
} else if (storedPreference === 'light') {
  classList.remove(schemeClasses);
  classList.add('ksv-scheme--light');
} else {
  classList.remove(schemeClasses);
  classList.add('ksv-scheme--auto');
}
`.replace(/\n|\s{2,}/g, '');

/** @deprecated use `nextRootScript` instead. */
const mediaQueryOnlyColorSchemeScaleScript = /*#__PURE__*/jsx("script", {
  dangerouslySetInnerHTML: {
    __html: script
  }
});
const nextRootScript = /*#__PURE__*/jsx("script", {
  dangerouslySetInnerHTML: {
    __html: script
  }
});

export { mediaQueryOnlyColorSchemeScaleScript, nextRootScript };
